README.md
<div align="center">
  <a href="https://ttytoolkit.org"><img width="130" src="https://github.com/piotrmurach/tty/raw/master/images/tty.png" alt="TTY Toolkit logo"/></a>
</div>

# TTY::Screen

[![Gem Version](https://badge.fury.io/rb/tty-screen.svg)][gem]
[![Actions CI](https://github.com/piotrmurach/tty-screen/actions/workflows/ci.yml/badge.svg)][gh_actions_ci]
[![Build status](https://ci.appveyor.com/api/projects/status/myjv8kahk1iwrlha?svg=true)][appveyor]
[![Code Climate](https://codeclimate.com/github/piotrmurach/tty-screen/badges/gpa.svg)][codeclimate]
[![Coverage Status](https://coveralls.io/repos/piotrmurach/tty-screen/badge.svg)][coverage]

[gitter]: https://gitter.im/piotrmurach/tty
[gem]: http://badge.fury.io/rb/tty-screen
[gh_actions_ci]: https://github.com/piotrmurach/tty-screen/actions/workflows/ci.yml
[appveyor]: https://ci.appveyor.com/project/piotrmurach/tty-screen
[codeclimate]: https://codeclimate.com/github/piotrmurach/tty-screen
[coverage]: https://coveralls.io/r/piotrmurach/tty-screen

> Terminal screen size detection which works on Linux, OS X and Windows/Cygwin platforms and supports MRI, JRuby, TruffleRuby and Rubinius interpreters.

**TTY::Screen** provides independent terminal screen size detection component for [TTY](https://github.com/piotrmurach/tty) toolkit.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'tty-screen'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install tty-screen

## 1. Usage

**TTY::Screen** allows you to detect terminal screen size by calling `size` method which returns [height, width] tuple.

```ruby
TTY::Screen.size     # => [51, 280]
```

To read terminal width do:

```ruby
TTY::Screen.width    # => 280
TTY::Screen.columns  # => 280
TTY::Screen.cols     # => 280
```

Similarly, to read terminal height do:

```ruby
TTY::Screen.height   # => 51
TTY::Screen.rows     # => 51
TTY::Screen.lines    # => 51
```

## Contributing

1. Fork it ( https://github.com/piotrmurach/tty-screen/fork )
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create a new Pull Request

This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [Contributor Covenant](http://contributor-covenant.org) code of conduct.

## Copyright

Copyright (c) 2014 Piotr Murach. See LICENSE for further details.


CHANGELOG.md
# Change log

## [v0.8.2] - unreleased

### Changed
* Change the ioctl system call support check to use the output method
* Change size from ioctl to extract window buffer format and length to constants
* Change size from ioctl to allocate and read window buffer with fewer objects
* Change size detection from tput to check TTY on the output method
* Change size detection from stty to check TTY on the output method
* Change size detection from the environment variables to use the env method
* Change size from the ansicon environment variable to use the env method
* Change size from ansicon to use regexp last match to lookup pattern matches
* Change size from ansicon to allocate window dimensions with fewer objects

### Fixed
* Fix ioctl call test to stub terminal size encoding for big-endian systems
* Fix to skip terminal size detection from readline on non-TTY output
* Fix size detection from IO to skip loading io-console on non-TTY output
* Fix size detection from tput to check for non-zero columns instead of lines
* Fix JRuby and Windows system detection to check Ruby configuration only once

## [v0.8.1] - 2020-07-17

### Fixed
* Fix name resolution with TTY::File by Alexey Nikitin (@tank-bohr)

## [v0.8.0] - 2020-05-28

### Added
* Add #windows? platform check
* Add #command_exist? to see if an executable exists before running it
* Add performance tests

### Changed
* Change #jruby? method to hoist within module
* Change #size_from_win_api to check only on windows platform and
  hoist definition within module
* Change #size_from_java to hoist within module
* Change #size_from_ioctl to:
  * check solaris-like system
  * scan all streams to see if any attached to a terminal
  * hoist definition within module
* Change #size_from_io_console to perform check on JRuby as well
* Change #size_from_readline to attempt to load readline gem
* Change #run_command to execute command directly without sub shell or temp files

## [v0.7.1] - 2020-02-02

### Changed
* Change gemspec to add metadata, remove test artifacts and load version directly

## [v0.7.0] - 2019-05-19

### Changed
* Change gemspec to load files directly without using git
* Change to relax development dependencies

## [v0.6.5] - 2018-07-13

### Changed
* Change to namespace version file to allow for direct vendoring

## [v0.6.4] - 2017-12-22

### Fixed
* Fix to suppress stderr output from run_command by Tero Marttila(@SpComb)

## [v0.6.3] - 2017-11-22

### Changed
* Change #size_from_tput & #size_from_stty to capture generic IO and command execution errors to make the calls more robust

### Fixed
* Fix #size_from_ioctl to handle Errno errors and deal with Errno::EOPNOTSUPP

## [v0.6.2] - 2017-11-04

### Fixed
* Fix #size_from_java to provide size only for non-zero values
* Fix #size_from_ioctl to provide size only for non-zero values

## [v0.6.1] - 2017-10-29

### Fixed
* Fix #size_from_win_api to provide size if non zero to avoid [1,1] size

## [v0.6.0] - 2017-10-29

### Added
* Add #size_from_ioctl check for reading terminal size with Unix ioctl
* Add #size_from_java check for reading terminal size from Java on JRuby
* Add #size_from_win_api check for reading terminal size from Windows C API

### Changed
* Change TTY::Screen to a module without any state
* Change to prefix all checks with `size` keyword
* Change gemspec to require ruby >= 2.0.0
* Remove #try_io_console and inline with io-console check
* Remove #default_size and replace with constant
* Remove TTY::Screen::Size class

## [v0.5.1] - 2017-10-26

### Changed
* Change #from_io_console to return nil when no size present
* Change #run_command to silently ignore any errors

### Fixed
* Fix #from_readline check to prevent from failing on missing api call
* Fix #from_stty to only extract size when stty command returns output
* Fix #run_command to correctly capture command output and fix #from_tput check

## [v0.5.0] - 2016-01-03

### Changed
* Change size to accept environment as input
* Remove Color detection, available as tty-color gem dependency

## [v0.4.3] - 2015-11-01

### Added
* Add NoValue to Color class to mark failure of reading color value

### Changed
* Change Color class supports? to recognize lack of color value

### Fixed
* Fix issue with #from_curses method and remove ensure block

## [v0.4.2] - 2015-10-31

### Changed
* Change visibility of output to prevent warnings

## [v0.4.1] - 2015-10-31

### Changed
* Change to switch off verbose mode by default

## [v0.4.0] - 2015-09-12

### Added
* Add terminal color support detection

## [v0.3.0] - 2015-09-11

### Fixed
* Fix bug loading standard library

## [v0.2.0] - 2015-05-11

### Changed
* Change to stop memoization of screen class instance method

### Fixed
* Fix bug with screen detection from_io_console by @luxflux

[v0.8.2]: https://github.com/piotrmurach/tty-screen/compare/v0.8.1...v0.8.2
[v0.8.1]: https://github.com/piotrmurach/tty-screen/compare/v0.8.0...v0.8.1
[v0.8.0]: https://github.com/piotrmurach/tty-screen/compare/v0.7.1...v0.8.0
[v0.7.1]: https://github.com/piotrmurach/tty-screen/compare/v0.7.0...v0.7.1
[v0.7.0]: https://github.com/piotrmurach/tty-screen/compare/v0.6.5...v0.7.0
[v0.6.5]: https://github.com/piotrmurach/tty-screen/compare/v0.6.4...v0.6.5
[v0.6.4]: https://github.com/piotrmurach/tty-screen/compare/v0.6.3...v0.6.4
[v0.6.3]: https://github.com/piotrmurach/tty-screen/compare/v0.6.2...v0.6.3
[v0.6.2]: https://github.com/piotrmurach/tty-screen/compare/v0.6.1...v0.6.2
[v0.6.1]: https://github.com/piotrmurach/tty-screen/compare/v0.6.0...v0.6.1
[v0.6.0]: https://github.com/piotrmurach/tty-screen/compare/v0.5.1...v0.6.0
[v0.5.1]: https://github.com/piotrmurach/tty-screen/compare/v0.5.0...v0.5.1
[v0.5.0]: https://github.com/piotrmurach/tty-screen/compare/v0.4.3...v0.5.0
[v0.4.3]: https://github.com/piotrmurach/tty-screen/compare/v0.4.2...v0.4.3
[v0.4.2]: https://github.com/piotrmurach/tty-screen/compare/v0.4.1...v0.4.2

spec/spec_helper.rb
# frozen_string_literal: true

if ENV["COVERAGE"] == "true"
  require "simplecov"
  require "coveralls"

  SimpleCov.formatter = SimpleCov::Formatter::MultiFormatter.new([
    SimpleCov::Formatter::HTMLFormatter,
    Coveralls::SimpleCov::Formatter
  ])

  SimpleCov.start do
    command_name "spec"
    add_filter "spec"
    enable_coverage :branch
  end
end

module Helpers
  # Replace standard streams
  #
  # @example
  #   replace_standard_streams(StringIO.new) do
  #     ...
  #   end
  #
  # @param [StringIO] output
  #   the output to replace standard streams with
  #
  # @return [void]
  #
  # @api public
  def replace_standard_streams(output)
    original_streams = [$stdin, $stdout, $stderr]
    $stdin, $stdout, $stderr = output, output, output
    yield
  ensure
    $stdin, $stdout, $stderr = *original_streams
  end

  # Undefine a constant
  #
  # @example
  #   undefine_const(:Readline) do
  #     ...
  #   end
  #
  # @param [String, Symbol] name
  #   the constant name
  #
  # @return [void]
  #
  # @api public
  def undefine_const(name)
    if Object.const_defined?(name)
      const = Object.send(:remove_const, name)
    end
    yield
  ensure
    Object.const_set(name, const) if const
  end
end

require "tty-screen"

RSpec.configure do |config|
  config.include(Helpers)

  config.expect_with :rspec do |expectations|
    expectations.include_chain_clauses_in_custom_matcher_descriptions = true
  end

  config.mock_with :rspec do |mocks|
    mocks.verify_partial_doubles = true
  end

  # Limits the available syntax to the non-monkey patched syntax
  # that is recommended.
  config.disable_monkey_patching!

  # This setting enables warnings. It's recommended, but in some cases may
  # be too noisy due to issues in dependencies.
  config.warnings = true

  if config.files_to_run.one?
    config.default_formatter = "doc"
  end

  config.profile_examples = 2

  config.order = :random

  Kernel.srand config.seed
end


spec/unit/new_spec.rb
RSpec.describe TTY::Screen, '.new' do

  it "initializses size with defaults" do
    allow(TTY::Screen::Size).to receive(:new)
    TTY::Screen.new
    expect(TTY::Screen::Size).to have_received(:new).
      with(ENV, {output: $stderr, verbose: false})
  end

  it "initializes size with values" do
    allow(TTY::Screen::Size).to receive(:new)

    TTY::Screen.new(output: :output, verbose: true)

    expect(TTY::Screen::Size).to have_received(:new).
      with(ENV, output: :output, verbose: true)
  end

  it "delegates size call" do
    size_instance = spy(:size)
    allow(TTY::Screen::Size).to receive(:new).and_return(size_instance)

    screen = described_class.new
    screen.size

    expect(size_instance).to have_received(:size)
  end

  it "calls size" do
    size_instance = double(:size, size: [51, 280])
    allow(TTY::Screen::Size).to receive(:new).and_return(size_instance)

    expect(TTY::Screen.size).to eq([51, 280])
    expect(TTY::Screen::Size).to have_received(:new).once
  end

  it "calls width" do
    size_instance = double(:size, size: [51, 280])
    allow(TTY::Screen::Size).to receive(:new).and_return(size_instance)

    expect(TTY::Screen.width).to eq(280)
    expect(TTY::Screen::Size).to have_received(:new).once
  end

  it "calls height" do
    size_instance = double(:size, size: [51, 280])
    allow(TTY::Screen::Size).to receive(:new).and_return(size_instance)
    expect(TTY::Screen.height).to eq(51)
  end
end


spec/perf/size_spec.rb
# frozen_string_literal: true

require "io/console"
require "rspec-benchmark"

RSpec.describe TTY::Screen, ".size" do
  include RSpec::Benchmark::Matchers

  it "detectes size 13x slower than io-console" do
    expect {
      TTY::Screen.size
    }.to perform_slower_than {
      IO.console.winsize
    }.at_most(13).times
  end

  it "performs at least 28K i/s" do
    expect { TTY::Screen.size }.to perform_at_least(28_000).ips
  end

  it "allocates at most 10 objects" do
    expect { TTY::Screen.size }.to perform_allocation(10).objects
  end
end


lib/tty/screen.rb
# frozen_string_literal: true

begin
  require "rbconfig"
rescue LoadError
end

require_relative "screen/version"

module TTY
  # Used for detecting screen properties
  #
  # @api public
  module Screen
    # The Ruby configuration
    #
    # @return [Hash]
    #
    # @api private
    RUBY_CONFIG = defined?(::RbConfig) ? ::RbConfig::CONFIG : {}
    private_constant :RUBY_CONFIG

    # Helper to define private functions
    def self.private_module_function(name)
      module_function(name)
      private_class_method(name)
    end

    case RUBY_CONFIG["host_os"] || ::RUBY_PLATFORM
    when /mswin|msys|mingw|cygwin|bccwin|wince|emc/
      def windows?; true end
    else
      def windows?; false end
    end
    module_function :windows?

    case RUBY_CONFIG["ruby_install_name"] || ::RUBY_ENGINE
    when /jruby/
      def jruby?; true end
    else
      def jruby?; false end
    end
    module_function :jruby?

    # Default terminal size
    #
    # @api public
    DEFAULT_SIZE = [27, 80].freeze

    @env = ENV
    @output = $stderr

    class << self
      # Holds the environment variables
      # @api public
      attr_accessor :env

      # Specifies an output stream
      # @api public
      attr_accessor :output
    end

    # Get terminal rows and columns
    #
    # @return [Array[Integer, Integer]]
    #   return rows and columns
    #
    # @api public
    def size(verbose: false)
      size_from_java(verbose: verbose) ||
      size_from_win_api(verbose: verbose) ||
      size_from_ioctl ||
      size_from_io_console(verbose: verbose) ||
      size_from_readline(verbose: verbose) ||
      size_from_tput ||
      size_from_stty ||
      size_from_env ||
      size_from_ansicon ||
      size_from_default
    end
    module_function :size

    def width
      size[1]
    end
    module_function :width

    alias columns width
    alias cols width
    module_function :columns
    module_function :cols

    def height
      size[0]
    end
    module_function :height

    alias rows height
    alias lines height
    module_function :rows
    module_function :lines

    # Default size for the terminal
    #
    # @return [Array[Integer, Integer]]
    #
    # @api private
    def size_from_default
      DEFAULT_SIZE
    end
    module_function :size_from_default

    # Determine terminal size with a Windows native API
    #
    # @return [nil, Array[Integer, Integer]]
    #
    # @api private
    if windows?
      STDOUT_HANDLE = 0xFFFFFFF5

      def size_from_win_api(verbose: false)
        require "fiddle" unless defined?(Fiddle)

        kernel32 = Fiddle::Handle.new("kernel32")
        get_std_handle = Fiddle::Function.new(
          kernel32["GetStdHandle"], [-Fiddle::TYPE_INT], Fiddle::TYPE_INT)
        get_console_buffer_info = Fiddle::Function.new(
          kernel32["GetConsoleScreenBufferInfo"],
          [Fiddle::TYPE_LONG, Fiddle::TYPE_VOIDP], Fiddle::TYPE_INT)

        format = "SSSSSssssSS"
        buffer = ([0] * format.size).pack(format)
        stdout_handle = get_std_handle.(STDOUT_HANDLE)

        get_console_buffer_info.(stdout_handle, buffer)
        _, _, _, _, _, left, top, right, bottom, = buffer.unpack(format)
        size = [bottom - top + 1, right - left + 1]
        size if nonzero_column?(size[1] - 1)
      rescue LoadError
        warn "no native fiddle module found" if verbose
      rescue Fiddle::DLError
        # non windows platform or no kernel32 lib
      end
    else
      def size_from_win_api(verbose: false); nil end
    end
    module_function :size_from_win_api

    # Determine terminal size on jruby using native Java libs
    #
    # @return [nil, Array[Integer, Integer]]
    #
    # @api private
    if jruby?
      def size_from_java(verbose: false)
        require "java"

        java_import "jline.TerminalFactory"
        terminal = TerminalFactory.get
        size = [terminal.get_height, terminal.get_width]
        size if nonzero_column?(size[1])
      rescue
        warn "failed to import java terminal package" if verbose
      end
    else
      def size_from_java(verbose: false); nil end
    end
    module_function :size_from_java

    # Detect screen size by loading io/console lib
    #
    # On Windows io_console falls back to reading environment
    # variables. This means any user changes to the terminal
    # size won't be reflected in the runtime of the Ruby app.
    #
    # @return [nil, Array[Integer, Integer]]
    #
    # @api private
    def size_from_io_console(verbose: false)
      return unless output.tty?

      require "io/console" unless IO.method_defined?(:winsize)
      return unless output.respond_to?(:winsize)

      size = output.winsize
      size if nonzero_column?(size[1])
    rescue Errno::EOPNOTSUPP
      # no support for winsize on output
    rescue LoadError
      warn "no native io/console support or io-console gem" if verbose
    end
    module_function :size_from_io_console

    if !jruby? && output.respond_to?(:ioctl)
      TIOCGWINSZ = 0x5413 # linux
      TIOCGWINSZ_PPC = 0x40087468 # macos, freedbsd, netbsd, openbsd
      TIOCGWINSZ_SOL = 0x5468 # solaris

      # The ioctl window size buffer format
      #
      # @return [String]
      #
      # @api private
      TIOCGWINSZ_BUF_FMT = "SSSS"
      private_constant :TIOCGWINSZ_BUF_FMT

      # The ioctl window size buffer length
      #
      # @return [Integer]
      #
      # @api private
      TIOCGWINSZ_BUF_LEN = TIOCGWINSZ_BUF_FMT.length
      private_constant :TIOCGWINSZ_BUF_LEN

      # Read terminal size from Unix ioctl
      #
      # @return [nil, Array[Integer, Integer]]
      #
      # @api private
      def size_from_ioctl
        buffer = Array.new(TIOCGWINSZ_BUF_LEN, 0).pack(TIOCGWINSZ_BUF_FMT)

        if ioctl?(TIOCGWINSZ, buffer) ||
           ioctl?(TIOCGWINSZ_PPC, buffer) ||
           ioctl?(TIOCGWINSZ_SOL, buffer)

          rows, cols, = buffer.unpack(TIOCGWINSZ_BUF_FMT)
          [rows, cols] if nonzero_column?(cols)
        end
      end

      # Check if ioctl can be called and any of the streams is
      # attached to a terminal.
      #
      # @return [Boolean]
      #   True if any of the streams is attached to a terminal, false otherwise.
      #
      # @api private
      def ioctl?(control, buf)
        ($stdout.ioctl(control, buf) >= 0) ||
          ($stdin.ioctl(control, buf) >= 0) ||
          ($stderr.ioctl(control, buf) >= 0)
      rescue SystemCallError
        false
      end
      module_function :ioctl?
    else
      def size_from_ioctl; nil end
    end
    module_function :size_from_ioctl

    # Detect screen size using Readline
    #
    # @api private
    def size_from_readline(verbose: false)
      return unless output.tty?

      require "readline" unless defined?(::Readline)
      return unless ::Readline.respond_to?(:get_screen_size)

      size = ::Readline.get_screen_size
      size if nonzero_column?(size[1])
    rescue LoadError
      warn "no readline gem" if verbose
    rescue NotImplementedError
    end
    module_function :size_from_readline

    # Detect terminal size from tput utility
    #
    # @api private
    def size_from_tput
      return unless output.tty? && command_exist?("tput")

      lines = run_command("tput", "lines")
      return unless lines

      cols = run_command("tput", "cols")
      [lines.to_i, cols.to_i] if nonzero_column?(cols)
    end
    module_function :size_from_tput

    # Detect terminal size from stty utility
    #
    # @api private
    def size_from_stty
      return unless output.tty? && command_exist?("stty")

      out = run_command("stty", "size")
      return unless out

      size = out.split.map(&:to_i)
      size if nonzero_column?(size[1])
    end
    module_function :size_from_stty

    # Detect terminal size from environment
    #
    # After executing Ruby code if the user changes terminal
    # dimensions during code runtime, the code won't be notified,
    # and hence won't see the new dimensions reflected in its copy
    # of LINES and COLUMNS environment variables.
    #
    # @return [nil, Array[Integer, Integer]]
    #
    # @api private
    def size_from_env
      return unless env["COLUMNS"] =~ /^\d+$/

      size = [(env["LINES"] || env["ROWS"]).to_i, env["COLUMNS"].to_i]
      size if nonzero_column?(size[1])
    end
    module_function :size_from_env

    # Detect terminal size from Windows ANSICON
    #
    # @api private
    def size_from_ansicon
      return unless env["ANSICON"] =~ /\((.*)x(.*)\)/

      size = [::Regexp.last_match(2).to_i, ::Regexp.last_match(1).to_i]
      size if nonzero_column?(size[1])
    end
    module_function :size_from_ansicon

    # Check if command exists
    #
    # @return [Boolean]
    #
    # @api private
    def command_exist?(command)
      exts = env.fetch("PATHEXT", "").split(::File::PATH_SEPARATOR)
      env.fetch("PATH", "").split(::File::PATH_SEPARATOR).any? do |dir|
        file = ::File.join(dir, command)
        ::File.exist?(file) ||
          exts.any? { |ext| ::File.exist?("#{file}#{ext}") }
      end
    end
    private_module_function :command_exist?

    # Runs command silently capturing the output
    #
    # @api private
    def run_command(*args)
      %x(#{args.join(" ")})
    rescue IOError, SystemCallError
      nil
    end
    private_module_function :run_command

    # Check if number is non zero
    #
    # return [Boolean]
    #
    # @api private
    def nonzero_column?(column)
      column.to_i > 0
    end
    private_module_function :nonzero_column?
  end # Screen
end # TTY


lib/tty/screen/color.rb
# coding: utf-8

module TTY
  class Screen
    class Color
      NoValue = Module.new

      # Initialize color support
      #
      # @api public
      def initialize(options = {})
        @output  = options.fetch(:output) { $stderr }
        @verbose = options.fetch(:verbose) { false }
      end

      # Detect if terminal supports color
      #
      # @return [Boolean]
      #   true when terminal supports color, false otherwise
      #
      # @api public
      def supports?
        return false unless tty?

        value = false
        %w(from_curses from_tput from_term from_env).each do |from_check|
          break if (value = public_send(from_check)) != NoValue
        end
        return false if value == NoValue
        value
      end

      # Attempt to load curses to check color support
      #
      # @return [Boolean]
      #
      # @api private
      def from_curses(curses_class = nil)
        require 'curses'

        if defined?(Curses)
          curses_class ||= Curses
          curses_class.init_screen
          has_color = curses_class.has_colors?
          curses_class.close_screen
          has_color
        else
          NoValue
        end
      rescue LoadError
        warn 'no native curses support' if @verbose
        NoValue
      end

      # Shell out to tput to check color support
      #
      # @api private
      def from_tput
        %x(tput colors 2>/dev/null).to_i > 2
      rescue Errno::ENOENT
        NoValue
      end

      # Inspect environment $TERM variable for color support
      #
      # @api private
      def from_term
        if ENV['TERM'] == 'dumb'
          false
        elsif ENV['TERM'] =~ /^screen|^xterm|^vt100|color|ansi|cygwin|linux/i
          true
        else NoValue
        end
      end

      def from_env
        ENV.include?('COLORTERM')
      end

      attr_reader :output

      def tty?
        output.tty?
      end
    end # Color
  end # Screen
end # TTY


lib/tty/screen/size.rb
# frozen_string_literal: true

module TTY
  class Screen
    class Size
      # Default terminal size
      #
      # @api public
      DEFAULT_SIZE = [27, 80].freeze

      # Initialize terminal size detection
      #
      # @api public
      def initialize(env, options = {})
        @env     = env
        @output  = options.fetch(:output) { $stderr }
        @verbose = options.fetch(:verbose) { false }
      end

      # Get terminal rows and columns
      #
      # @return [Array[Integer, Integer]]
      #   return rows & columns
      #
      # @api public
      def size
        size = from_java
        size ||= from_win_api
        size ||= from_ioctl
        size ||= from_io_console
        size ||= from_readline
        size ||= from_tput
        size ||= from_stty
        size ||= from_env
        size ||= from_ansicon
        size ||  DEFAULT_SIZE
      end

      STDOUT_HANDLE = 0xFFFFFFF5

      # Determine terminal size with a Windows native API
      #
      # @return [nil, Array[Integer, Integer]]
      #
      # @api private
      def from_win_api
        require 'fiddle'

        kernel32 = Fiddle::Handle.new('kernel32')
        get_std_handle = Fiddle::Function.new(kernel32['GetStdHandle'],
                          [-Fiddle::TYPE_INT], Fiddle::TYPE_INT)
        get_console_buffer_info = Fiddle::Function.new(
          kernel32['GetConsoleScreenBufferInfo'],
          [Fiddle::TYPE_LONG, Fiddle::TYPE_VOIDP], Fiddle::TYPE_INT)

        format        = 'SSSSSssssSS'
        buffer        = ([0] * format.size).pack(format)
        stdout_handle = get_std_handle.(STDOUT_HANDLE)

        get_console_buffer_info.(stdout_handle, buffer)
        _, _, _, _, _, left, top, right, bottom, = buffer.unpack(format)
        [bottom - top + 1, right - left + 1]
      rescue LoadError
        warn 'no native fiddle module found' if @verbose
      rescue Fiddle::DLError
        # non windows platform or no kernel32 lib
      end

      # Determine terminal size on jruby using native Java libs
      #
      # @return [nil, Array[Integer, Integer]]
      #
      # @api private
      def from_java
        return unless jruby?
        require 'java'
        java_import 'jline.TerminalFactory'
        terminal = TerminalFactory.get
        [terminal.get_height, terminal.get_width]
      rescue
        warn 'failed to import java terminal package' if @verbose
      end

      # Detect screen size by loading io/console lib
      #
      # On Windows io_console falls back to reading environment
      # variables. This means any user changes to the terminal
      # size won't be reflected in the runtime of the Ruby app.
      #
      # @return [nil, Array[Integer, Integer]]
      #
      # @api private
      def from_io_console
        return if jruby?
        require 'io/console'

        begin
          if output.tty? && IO.method_defined?(:winsize)
            size = output.winsize
            size if nonzero_column?(size[1])
          end
        rescue Errno::EOPNOTSUPP
          # no support for winsize on output
        end
      rescue LoadError
        warn 'no native io/console support or io-console gem' if @verbose
      end

      TIOCGWINSZ = 0x5413
      TIOCGWINSZ_PPC = 0x40087468

      # Read terminal size from Unix ioctl
      #
      # @return [nil, Array[Integer, Integer]]
      #
      # @api private
      def from_ioctl
        return if jruby?
        return unless output.respond_to?(:ioctl)

        format = 'SSSS'
        buffer = ([0] * format.size).pack(format)
        if ioctl?(TIOCGWINSZ, buffer) || ioctl?(TIOCGWINSZ_PPC, buffer)
          rows, cols, = buffer.unpack(format)[0..1]
          return [rows, cols]
        end
      end

      def ioctl?(control, buf)
        output.ioctl(control, buf) >= 0
      rescue Errno::ENOTTY
        # wrong processor architecture
        false
      rescue Errno::EINVAL
        # ioctl failed to recognise processor type(not Intel)
        false
      end

      # Detect screen size using Readline
      #
      # @api private
      def from_readline
        if defined?(Readline) && Readline.respond_to?(:get_screen_size)
          size = Readline.get_screen_size
          size if nonzero_column?(size[1])
        end
      rescue NotImplementedError
      end

      # Detect terminal size from tput utility
      #
      # @api private
      def from_tput
        return unless output.tty?
        lines = run_command('tput', 'lines').to_i
        cols  = run_command('tput', 'cols').to_i
        [lines, cols] if nonzero_column?(lines)
      rescue Errno::ENOENT
      end

      # Detect terminal size from stty utility
      #
      # @api private
      def from_stty
        return unless output.tty?
        out = run_command('stty', 'size')
        return unless out
        size = out.split.map(&:to_i)
        size if nonzero_column?(size[1])
      rescue Errno::ENOENT
      end

      # Detect terminal size from environment
      #
      # After executing Ruby code if the user changes terminal
      # dimensions during code runtime, the code won't be notified,
      # and hence won't see the new dimensions reflected in its copy
      # of LINES and COLUMNS environment variables.
      #
      # @return [nil, Array[Integer, Integer]]
      #
      # @api private
      def from_env
        return unless @env['COLUMNS'] =~ /^\d+$/
        size = [(@env['LINES'] || @env['ROWS']).to_i, @env['COLUMNS'].to_i]
        size if nonzero_column?(size[1])
      end

      # Detect terminal size from Windows ANSICON
      #
      # @api private
      def from_ansicon
        return unless @env['ANSICON'] =~ /\((.*)x(.*)\)/
        size = [$2, $1].map(&:to_i)
        size if nonzero_column?(size[1])
      end

      # Specifies an output stream object
      #
      # @api public
      attr_reader :output

      private

      # Runs command silently capturing the output
      #
      # @api private
      def run_command(*args)
        require 'tempfile'
        out = Tempfile.new('tty-screen')
        result = system(*args, out: out.path)
        return if result.nil?
        out.rewind
        out.read
      ensure
        out.close if out
      end

      # Check if number is non zero
      #
      # return [Boolean]
      #
      # @api private
      def nonzero_column?(column)
        column.to_i > 0
      end

      def jruby?
        RbConfig::CONFIG['ruby_install_name'] == 'jruby'
      end
    end # Size
  end # Screen
end # TTY
